src/citrine/resources/ara_definition.py

```diff
@@ -124,6 +126,74 @@ def add_columns(self, *,
+                            project,
+                            quantity_dimension: IngredientQuantityDimension,
+                            scope: str = 'id'
+                            ):
+        """[ALPHA] Add variables and columns for all of the possible ingredients in a process.
+
+        For each allowed ingredient name in the process template there is a column for the if of
+        the ingredient and a column for the quantity of the ingredient. If the quantities are
+        given in absolute amounts then there is also a column for units.
+
+        Parameters
+        ------------
+        process_template: LinkByUID
+            scope and id of a registered process template
+        project: Project
+            a project that has access to the process template
+        quantity_dimension: IngredientQuantityDimension
+            the dimension in which to report ingredient quantities
+        scope: Optional[str]
+            the scope for which to get ingredient ids (default is Citrine scope, 'id')
+
+        """
+        quantity_dimension_display = {
+            IngredientQuantityDimension.ABSOLUTE: "absolute quantity",
+            IngredientQuantityDimension.MASS: "mass fraction",
+            IngredientQuantityDimension.VOLUME: "volume fraction",
+            IngredientQuantityDimension.NUMBER: "number fraction"
+        }
+        process: ProcessTemplate = project.process_templates.get(
+            uid=process_template.id, scope=process_template.scope)
+        new_variables = []
+        new_columns = []
+        for name in process.allowed_names:
+            identifier_variable = IngredientIdentifierByProcessTemplateAndName(
+                name="{} of ingredient {} in process {}".format(scope, name, process.name),
+                headers=[process.name, name + ' ' + scope],
+                process_template=process_template,
+                ingredient_name=name,
+                scope=scope
+            )
+            quantity_variable = IngredientQuantityByProcessAndName(
+                name="{} of ingredient {} in process {}".format(
+                    quantity_dimension_display[quantity_dimension], name, process.name),
+                headers=[process.name, name, quantity_dimension_display[quantity_dimension]],
+                process_template=process_template,
+                ingredient_name=name,
+                quantity_dimension=quantity_dimension
+            )
+
+            if identifier_variable.name not in [var.name for var in self.variables]:
+                new_variables.append(identifier_variable)
+                new_columns.append(IdentityColumn(data_source=identifier_variable.name))
+            new_variables.append(quantity_variable)
+            new_columns.append(MeanColumn(data_source=quantity_variable.name))
+            if quantity_dimension == IngredientQuantityDimension.ABSOLUTE:
+                new_columns.append(OriginalUnitsColumn(data_source=quantity_variable.name))
```

Comment:
We should consider if we want to create a mechanism for `automatically cast this into SI` rather than create the column.

Reply:
But even if we cast to SI, wouldn't it still be better from a user perspective to see the units next to the quantity?

Reply:
The question is if we feel confident that the user won't mix volume and mass in one column.  I'm not certain I feel confident on that, so your approach makes sense.  I'm just generally very against mixing units within one column.

Reply:
Are you imagining a situation in which the user adds ingredients in an ABSOLUTE basis and some ingredient amounts are given in `mg` and some are in `ml`? And are you suggesting breaking that out into 2 columns, one for the values in units of mass and one for the values in units of volume?

Reply:
That's kinda where my thoughts are.  We can save this for another PR.

Reply:
This is a case where I would _hope_ the user uses the same quantity type for that ingredient across all processes (and this should be the case most of the time), but I would warn the user if that were not the case

Reply:
There are 2 distinct issues.
1. The user reports quantities with different dimensions (e.g. some things in mass fraction and others in volume fraction). I think this is what you are referring to Nils. In this case I think they would get a table with a bunch of NaN values in it, but no warning. I believe a warning would have to be generated in Faraday.
2. The user reports all quantities in the ABSOLUTE dimension but the units are of different types (e.g., mg for some and ml for others). I think that's less of a problem. Sometimes different ingredients are more naturally expressed in different quantities. Our formulations stack and optimizer should be able to handle it  (the "share in" descriptors won't be easily interpretable by a human, but it should still work)